2.0 refactoring

andrew-codes opened the pull request:
- refactoring of core library
- introduction of high order functions (connectors) to provide sdk with AJAX implementations

walkerrandolphsmith commented on src/createMeta.js:
```diff
@@ -0,0 +1,60 @@
+        },
+
+        queryDefinition(assetType) {
+            invariant(assetType, `Error: there was no \`assetType\` provided to queryDefinition`);
```
v1.meta does not require an assetType and if ignored will return the entire definition of Meta.

JogoShugh commented on src/connectors/jqueryConnector.js:
```diff
@@ -0,0 +1,17 @@
+    })),
+    (url, data, headers) => new Promise((resolve, reject) => $.ajax(url, {
+        method: 'GET',
+        data: JSON.stringify(data),
```
Does jquery not take care of query string Parameters or is that what it uses data for here?

andrew-codes replied:
@JogoShugh I believe jQuery handles the data as a query string with the method is `'GET'`

JogoShugh commented on src/createMeta.js:
```diff
@@ -0,0 +1,60 @@
+const createHeaderObj = (token) => ({
+    Accept: 'application/json',
+    'Content-Type': 'application/json',
+    Authorization: `Basic ${token}`
```
Tokens are prefixed with Bearer not basic.

walkerrandolphsmith replied:
I hope there is not a hard requirement on providing an access token. Some consumers (spigot) need to use basic authentication to perform intentions by a specific Member.

andrew-codes replied:
@walkerrandolphsmith there is not, you can use an access token or username/password via `withCreds('username', 'password')`

andrew-codes commented:
@JogoShugh and @walkerrandolphsmith, I addressed the concerns you had about:
* authorization bearer versus basic
* querying meta definitions for all of meta instead of just for a single asset type

Any other concerns?

walkerrandolphsmith commented:
Yes I have an issue with requiring an access token. Spigot requires actions be made on the behalf of specific users. Members can generate personal access tokens which could satisfy this case, however spigot is an automation tool and generating access tokens is not an automated process leaving spigot to require user interaction to work. If you allow basic authentication or access tokens then this will not be an issue for consumers like spigot. @andrew-codes

JogoShugh commented on src/transformDataToAsset.js:
```diff
@@ -1,34 +1,42 @@
-                name: key,
-                value: attributeData.map(reduceRelationalAttributes)
+            return {
+                ...output,
```
Wow...so I'm assuming "output" is an object here, and ...output in this context means
expand all its properties into this new literal obejct?
